### PR TITLE
Add delegator

### LINKED ISSUE #781: Add Delegator support for elastic

### PR DESCRIPTION
**Pull request by sukantoraymond**
closes #781

### CODE REVIEW

File: cmd/subnetcmd/join.go [hunk 1]
```diff
@@ -490,11 +493,13 @@ func promptStakeAmount(subnetName string) (uint64, error) {
 		return 0, err
 	}
 	maxValidatorStake := fmt.Sprintf("Maximum Validator Stake (%d)", esc.MaxValidatorStake)
-	customWeight := "Custom (Has to be between minValidatorStake and maxValidatorStake defined during elastic subnet transformation)"
+	customWeight := fmt.Sprintf("Custom (Has to be between minValidatorStake (%d) and maxValidatorStake (%d) defined during elastic subnet transformation)", esc.MinValidatorStake, esc.MaxValidatorStake)
+	if !isValidator {
+		customWeight = fmt.Sprintf("Custom (Has to be between minDelegatorStake (%d) and maxValidatorStake (%d) defined during elastic subnet transformation)", esc.MinDelegatorStake, esc.MaxValidatorStake)
```
Comment (felipemadero) [inline cmd/subnetcmd/join.go:498]: why not MaxDelegatorStake?

File: cmd/subnetcmd/addDelegator.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+		return err
+	}
+	endTime := start.Add(stakeDuration)
+	ux.Logger.PrintToUser("Inputs complete, issuing transaction addPermissionlessDelegatorTx...")
+	ux.Logger.PrintToUser("")
+
+	assetID := sc.ElasticSubnet[models.Local.String()].AssetID
+	testKey := genesis.EWOQKey
+	keyChain := secp256k1fx.NewKeychain(testKey)
+	subnetID := sc.Networks[models.Local.String()].SubnetID
```
Comment (felipemadero) [inline cmd/subnetcmd/addDelegator.go:141]: can you avoid using models.Local and instead use network

File: tests/e2e/testcases/subnet/local/suite.go [hunk 3]
```diff
@@ -204,7 +204,7 @@ var _ = ginkgo.Describe("[Local Subnet]", ginkgo.Ordered, func() {
 		_, err = commands.RemoveValidator(subnetName, nodeIDs[1])
 		gomega.Expect(err).Should(gomega.BeNil())
 
-		_, err = commands.AddPermissionlessValidator(subnetName, nodeIDs[1], stakeAmount, stakeDuration)
+		_, err = commands.AddPermissionlessValidator(subnetName, nodeIDs[1], "200000", stakeDuration)
```
Comment (felipemadero) [inline tests/e2e/testcases/subnet/local/suite.go:207]: are you adding a test for add delegator?

Comment (felipemadero) [inline tests/e2e/testcases/subnet/local/suite.go:207]: why avoiding a constant here?

File: cmd/subnetcmd/join.go [hunk 4]
```diff
@@ -453,23 +453,26 @@ func promptNodeIDToAdd(subnetID ids.ID) (ids.NodeID, error) {
...
 		for _, localNodeID := range defaultLocalNetworkNodeIDs {
 			nodeIDFound := false
-			for _, v := range validators {
-				if v.NodeID.String() == localNodeID {
-					nodeIDFound = true
-					break
+			if isValidator {
+				for _, v := range validators {
+					if v.NodeID.String() == localNodeID {
+						nodeIDFound = true
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/subnetcmd/join.go:460]: yes, but it seems more readable with continue, and more direct

File: cmd/subnetcmd/join.go [hunk 5]
```diff
@@ -453,23 +453,26 @@ func promptNodeIDToAdd(subnetID ids.ID) (ids.NodeID, error) {
 		}
 		// construct list of validators to choose from
 		var validatorList []string
-		fmt.Printf("defaultLocalNetworkNodeIDs %s \n", defaultLocalNetworkNodeIDs)
-
 		for _, localNodeID := range defaultLocalNetworkNodeIDs {
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/subnetcmd/join.go:455]: You should get cluter and with it get the actual list of nodes

**Review by felipemadero — COMMENTED** [review]
It is ok to me after addrssing remaining conversations, and integrating https://github.com/ava-labs/avalanche-cli/pull/821/files 
I think the reason for the e2e problem was related to validation window problem, not overdelegation problem.
The above PR add a delegation case for e2e. Also made https://github.com/ava-labs/avalanchego-internal/pull/2836/files to avoid confusing over delegation problems with staking window problems.

File: cmd/subnetcmd/addDelegator.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+		return errors.New("addPermissionlessDelegator is not yet supported on Mainnet")
+	}
+
+	if outputTxPath != "" {
+		if _, err := os.Stat(outputTxPath); err == nil {
+			return fmt.Errorf("outputTxPath %q already exists", outputTxPath)
+		}
+	}
+
+	if len(ledgerAddresses) > 0 {
```
Comment (dasconnor) [inline cmd/subnetcmd/addDelegator.go:106]: we don't usually enable ledgers for local development. I don't think you need to explicitly remove this, but just letting you know

File: cmd/subnetcmd/join.go [hunk 7]
```diff
@@ -444,32 +444,64 @@ func checkIsValidating(subnetID ids.ID, nodeID ids.NodeID, pClient platformvm.Cl
 	return false, nil
 }
 
-func promptNodeIDToAdd(subnetID ids.ID) (ids.NodeID, error) {
+func getLocalNetworkIDs() ([]string, error) {
```
Comment (dasconnor) [inline cmd/subnetcmd/join.go:447]: Should be `NodeIDs`? I thought we already had a method for this?

**Review by dasconnor — COMMENTED** [review]
Looks really close!

**Review by felipemadero — APPROVED** [review]
LGTM. Should anyway wait also for @dasconnor approval

**Review by dasconnor — APPROVED** [review]

**Pull request merged**